Source/Core/Core/Scripting/ScriptUtilities.cpp

```diff
@@ -0,0 +1,764 @@
+                                           new_unique_script_identifier, new_script_filename,
+                                           new_print_callback_func, new_script_end_callback_func));
+}
+
+void PushScriptStopQueueEvent(const ScriptQueueEventTypes event_type, const int script_identifier)
+{
+  std::lock_guard<std::mutex> lock(initialization_and_destruction_lock);
+  is_scripting_core_initialized = true;
+  script_events.push_back(ScriptQueueEvent(event_type, script_identifier, "", nullptr, nullptr));
+}
+
+void ProcessScriptQueueEvents()
+{
+  // Core::CPUThreadGuard lock(Core::System::GetInstance());
+  std::lock_guard<std::mutex> lock(initialization_and_destruction_lock);
+  if (script_events.size() == 0)
+    return;
+
+  // This loop deletes all duplicate events from the queue
+  for (unsigned long long i = 0; i < script_events.size() - 1; ++i)
+  {
+    for (unsigned long long j = i + 1; j < script_events.size(); ++j)
+    {
+      if (script_events[i] == script_events[j])
+      {
+        script_events.erase(script_events.begin() + j);
+        --j;
+      }
+    }
+  }
+
+  // Now, removing all pairs of start and stop script events that have the same unique script
+  // identifier
+  for (unsigned long long i = 0; i < script_events.size() - 1; ++i)
+  {
+    if (script_events[i].event_type == ScriptQueueEventTypes::CreateScriptFromUI)
+    {
+      int current_script_identifier = script_events[i].unique_script_identifier;
+      for (unsigned long long j = i + 1; j < script_events.size(); ++j)
+      {
+        if (script_events[j].unique_script_identifier == current_script_identifier &&
+            script_events[j].event_type == ScriptQueueEventTypes::StopScriptFromUI)
+        {
+          script_events.erase(script_events.begin() + j);
+          script_events.erase(script_events.begin() + i);
+          break;
+        }
+      }
+    }
+  }
+
+  // Now, we begin processing the events in the queue in the order that they were added to the
+  // queue.
+  while (script_events.size() > 0)
+  {
+    ScriptQueueEvent current_event = script_events[0];
+    if (current_event.event_type == ScriptQueueEventTypes::CreateScriptFromUI)
+    {
+      InitializeScript(current_event.unique_script_identifier, current_event.script_filename,
+                       current_event.print_callback_func, current_event.script_end_callback_func);
+    }
+    else if (current_event.event_type == ScriptQueueEventTypes::StopScriptFromUI ||
+             current_event.event_type == ScriptQueueEventTypes::StopScriptFromScriptEndCallback)
+    {
+      StopScript(current_event.unique_script_identifier);
+    }
+    else
+    {
+      queuePanicAlertEvent("Error: Encountered unknown ScriptQueueEventType in "
+                           "ProcessScriptQueueEvents() function.");
+      return;
+    }
+    script_events.erase(script_events.begin());
+  }
+}
+
+bool StartScripts()
+{
+  std::lock_guard<std::mutex> lock(initialization_and_destruction_lock);
```

Comment:
If you're acquiring two mutexes in a row, you may want to do

```cpp
std::scoped_lock lock(initialization_and_destruction_lock, global_code_and_frame_callback_running_lock);
```

which functions like an in-place `std::lock`, preventing deadlocks